### PR TITLE
Switch to Next.js

### LINKED ISSUE #131: Flattening/restructuring the component directory
This is an issue that I will probably handle, but some of these component paths are a bit absurd, ex

Screen Shot 2021-05-03 at 5 48 56 PM

Some goals for me (as a short-term refactor):

* can we differentiate between pages and non-pages? is this too destructive?
* can we localize some shared components into one folder
* if a component is only ever a child of another component and is completely static, can we instead put it inside another class file?

### LINKED ISSUE #162: Scroll location on page change
When switching from one page on the website to another via the navbar or the logo button, the page opens at the same corresponding section (at the same location) instead of opening at the top

i.e. if you scroll to the bottom in the committees page and click the navbar or the main logo it will take you to the bottom of the corresponding page you clicked on, instead of to the top (which would be better behavior)

According to @mattxwang this probably has to do with some `e.preventDefaultScroll` shenanigans, and should be auto resolved on switching to next.js / gatsby - so just putting this up for posterity / as a reminder!

### PR DESCRIPTION
**Pull request by mattxwang**
This PR does the meat of our transition to Next.js. As a heads up, the files changed/diff is absolutely nasty; I imagine that there will be little bits and pieces that we'll have to address down the line.

## Summary

We transition the site to be Next SSG'd.

### CRA-to-Next transition

* swaps out toolchain to use Next
  * installs `next`, and relevant `eslint` plugins
  * removes `react-scripts`, `react-router-dom`, `react-router-hash-link`, and `@testing-library` equivalents
  * changes `npm` scripts to use `next` defaults
  * swaps out `node-sass` for `sass` (the reference dart implementation)
* uses Next's file-based routing
  * currently segments into six `pages`: `about`, `committee`, `events`, `index` (previously `Home`), `sponsors`, and `techgala`; no paths are broken
  * moves the rest of components (including `GM`) into `components`
* restructures SASS and CSS
  * moves all styles into `styles` folder, mirroring `pages` and `components` structure
  * collates a few styles
  * imports Font Awesome into the next bundle, instead of loading into the `index.html` entrypoint
* uses `next/image` when possible
  * in some cases, we import the image out of the `public` folder; this automatically sets `width` and `height`
  * in some other cases, manually sets `width` and `height` (typically for UI elements) or uses `layout="fill"` and retools CSS to properly deal with it
  * **in some cases, maintains ``. we were unable to fully transition over to `<Image />` (see below conversation)**
* for ease-of-transition, removes all tests (`jest`, `enzyme`, related things)

### Things that will need to be addressed later

* this PR explicitly doesn't address `.module` SASS, but we should do that in a follow-up
* ~there should be something we can do about fonts that doesn't involve loading it in the `<Head>` - let's look into that!~ possibly see @BryanPan342's comment below
* SEO, using `<Head>` properly, etc; since this isn't a regression, let's table adding this to each page in a separate PR
* more ergonomic image inputs - maybe through an `index.js` that collates together images, thanks @dtjanaka for the suggestion!
* remove `postcss@8` as a dev dependency, but still resolve this error: https://github.com/vercel/next.js/issues/24882
* refactor the `GM` component to derive all of its information from state; then, make a new wrapper page that passes in the relevant information from data files, manual configuration, etc.
* document the new CI/CD pipeline, document Next quirks (ex `next/image`)
* resolve issues with `next/image`

### Misc. code quality changes

* removes the `<Button>` component entirely, and changes the `button` selector to instead use the `.button` class.
  * this is because the component itself didn't do anything (it just passed on `props` to a `<button>`) and also arbitrarily restricted the semantic element
* changes CSS for the officers section of the about page to use flexbox instead of nested selectors + absolute/inline-based positioning
* changes the officers section text to say "Leadership" <- this is easily changeable
* fixes bug where `.page-content p` bleeds into the Committees page events arbitrarily
* fixes incorrect use of `ul` and `li` in `<Navbar>`
* slashes a non-trivial amount of completely unused CSS
* de-dupes `CommitteeSectionIntro`'s double-images for committee motifs

### Issues that still need to be addressed:

- [x] switching over to `next/image` to resolve build failures
- [x] properly using `next/image` to resolve layout issues - see `TODO` comments
- [x] font awesome icons are not sized properly
- [x] strip out the `Button` component, as it's not useful and causes issues with `Link`
- [x] fix the Navbar button (general_
- [x] fix the Navbar membership portal button on mobile
- [x] ~~this error: https://github.com/vercel/next.js/issues/24882~~ - stopgap is to have `postcss@8` as a devdep; moved to "deal with later"
- [x] resolve netlify build error (very simple, just change output folder)
- [x] ~~deal with the `GM` component -> this should probably be a component, and then have a dummy page wrap it for every new GM. deals with longevity very well~~ - moved to "deal with later"
- [x] fix about page first image diamond (now not centered properly)
- [x] fix image borders not appearing properly on `next/image` components
- [x] fix committee page nav images not being properly sized
- [x] the favicon is bugged

This PR also does some favours for us: it closes #162, (reasonably) closes #131 (for now). It also unblocks #110 and #116!

### CODE REVIEW

**Comment by BryanPan342** [comment]
@mattxwang lmk if u need help with the next.js migration!

Even tho next is awesome, sometimes the bugs are opaque AF and takes a bit of digging to fix!

**Comment by mattxwang** [comment]
> @mattxwang lmk if u need help with the next.js migration!
> 
> Even tho next is awesome, sometimes the bugs are opaque AF and takes a bit of digging to fix!

Thanks, appreciate it! I might pick your brain once I resolve all the low hanging fruit. Gotta say, the image conversion has been the largest pain point thus far, though it's also a huge code smell with our website in general!

**Comment by BryanPan342** [comment]
For fonts, usually we import them into our variables or global scss file!

https://github.com/UCLA-Creative-Labs/sunshine/blob/master/styles/_variables.module.scss

File: pages/_document.js
Comment (BryanPan342) [inline pages/_document.js:11]: You can link to `/images/logo.png`

Next.js takes ur public folder and treats it like root!

**Review by BryanPan342 — COMMENTED** [review]
Why are u using document? What part of the html/body tag do you really need to target?

File: pages/_document.js
Comment (BryanPan342) [reply pages/_document.js:11]: Yeah I think anything with a `public 
_url` you should be able to just remove! Also helps you from having to deal with that during production, which definitely won’t work with the base webpack plugins

Comment (mattxwang) [reply pages/_document.js:11]: Great catch! I'll resolve;

> Why are u using document? What part of the html/body tag do you really need to target?

The first-party next eslint plugin [throws an error](https://nextjs.org/docs/messages/no-page-custom-font) when you try to import a font (or in general, anything that should span every single page) in `_app.js`. Seems like you're supposed to slap all of that stuff in `_document.js`, but I could be totally off base here; still new to things!

Comment (BryanPan342) [reply pages/_document.js]: Awww I see.. for fonts check out the comment I made above.. it should remove that necessity! 

As for importing things for every page, what I like to do (and what I have seen elsewhere) is basically to have a Layout component (which you already have) that takes care of all your headers!

https://github.com/UCLA-Creative-Labs/tears/blob/main/components/Layout.tsx

You can even create different Layout types and have those be utilized throughout ur app!

**Review by advaithg — COMMENTED** [review]
I fixed the issues with the weird sizing of Next Images by making the attributes of the Image element `layout: "fill"` with `objectFit:"contain"` and wrapping them in a div which had the size we needed along with `position:relative`

This made sense cause the images on the website mostly had a fixed height but varying width, and Next didn't let me do that - so I had to do that in an outer div. I couldn't find a way to tell Next to "use default width"

Alignment was done using the `objectPosition` attribute

**Comment by mattxwang** [comment]
@BryanPan342, I'm curious do y'all host CL stuff on Vercel? Running into some issues with using `next/image` on Netlify (you can peep [the deploy preview](https://deploy-preview-178--jovial-pasteur-581b4a.netlify.app/) or check out the discussion https://github.com/vercel/next.js/issues/18356 or https://github.com/vercel/next.js/discussions/19065 for some context). No worries at all on timing though, you're doing a huge help for us already!

**Review by advaithg — COMMENTED** [review]
The current issue is that images fail to load - not sure exactly why.

@BryanPan342 do you know of a way to possibly fix this? (if you've experienced it before)

The most plausible reason we found is [this issue](https://github.com/vercel/next.js/issues/18356), but the only ways to fix it seem to be either switching to [next-optimized-images](https://github.com/cyrilwanner/next-optimized-images) or to vercel instead of using netlify.

**Comment by mattxwang** [comment]
> For fonts, usually we import them into our variables or global scss file!
> 
> https://github.com/UCLA-Creative-Labs/sunshine/blob/master/styles/_variables.module.scss

Never got back to this; I'm curious, does this delay the font load time? Since then we'll have to hit the CSS file first (fetch + parse), and then load the font itself (fetch + parse)?

**Comment by BryanPan342** [comment]
> Never got back to this; I'm curious, does this delay the font load time? Since then we'll have to hit the CSS file first (fetch + parse), and then load the font itself (fetch + parse)?

it does a little bit (I only really notice it when I'm really laggy), but I mean, it will load with the rest of your styling so I don't see a problem with it TBH

**Comment by BryanPan342** [comment]
> @BryanPan342, I'm curious do y'all host CL stuff on Vercel or Netlify? Running into some issues with using `next/image` on Netlify (you can peep [the deploy preview](https://deploy-preview-178--jovial-pasteur-581b4a.netlify.app/) or check out the discussion [vercel/next.js#18356](https://github.com/vercel/next.js/issues/18356) or [vercel/next.js#19065](https://github.com/vercel/next.js/discussions/19065) for some context). No worries at all on timing though, you're doing a huge help for us already!

we host everything on netlify! I haven't run into many problems with next/image BUT to be honest, I prefer using just the regular `image` tag! I find `next/image` abstracts a little too much for me personally.

**Comment by mattxwang** [comment]
> we host everything on netlify! I haven't run into many problems with next/image BUT to be honest, I prefer using just the regular `image` tag! I find `next/image` abstracts a little too much for me personally.

Hm okay thanks for the thought - I think we'll switch back to `` for now (since it moves gracefully), and try our hand at `<Image />` some other time (I also saw that there's some movement with the Essential Next.js plugin).

File: components/GM.js
Comment (dtjanaka) [inline components/GM.js:1]: Why is this file completely commented out?

File: components/TechGala/ProjectCard.js
Comment (dtjanaka) [inline components/TechGala/ProjectCard.js:9]: Maybe a dumb question, but it seems like height and width are now required props for images; what are the implications for responsiveness?

File: data/sponsors.js
Comment (dtjanaka) [inline data/sponsors.js:1]: maybe consider making an index.js file for the images so u can just do `import {blizzard, citadel, deloitee, etc.} from '../public/images/sponsors/';`

**Review by dtjanaka — APPROVED** [review]
nice

File: components/GM.js
Comment (advaithg) [reply components/GM.js:1]: I think GM needs to be refactored into an actual component in a dummy page, will probably be another issue!
@mattxwang should have more info

File: components/TechGala/ProjectCard.js
Comment (advaithg) [reply components/TechGala/ProjectCard.js:9]: The images automatically readjust themselves based on the size of the page, iirc. The height and width props are for the general case.

File: components/GM.js
Comment (mattxwang) [reply components/GM.js:1]: Yeah, just to follow up - since the page isn't live, didn't want to introduce more complexity into the change. Eventually, we should adjust this too.

A quick thought - I think the page for a GM (ex `gm-w21`) should wrap the existing `GM` component, and simply pass in props to the component. That way, the `GM` component itself doesn't need to rely on external data sources, which may be a better separation of concerns.

File: data/sponsors.js
Comment (mattxwang) [reply data/sponsors.js:1]: Great idea, let's add it to the backlog 😄

File: components/TechGala/ProjectCard.js
Comment (mattxwang) [reply components/TechGala/ProjectCard.js:9]: Yeah, I think moving forward @advaithg is going to be the SME for `next/image` - you may want to discuss how the component works (ex specified with/height versus `layout="fill"`, what's happening under the hood) with the team some time.

File: components/Layout.js
Comment (evanzhong) [inline components/Layout.js:17]: This is so clean, really nice

File: components/Navbar.js
Comment (evanzhong) [inline components/Navbar.js:67]: Similarly, remove if possible before merging

File: components/Navbar.js
Comment (evanzhong) [inline components/Navbar.js:8]: As this is written right now, this is kind of a hack, but it's just a result of how we designed the `Navbar` component

I'm curious what y'alls thoughts are on making this more friendly to page inactivation. We could easily store the `ul` information in a json object and have the Navbar be constructed from it. I think this would also benefit us by making the code DRYer.

File: pages/events.js
Comment (evanzhong) [inline pages/events.js:33]: Is the choice to remove the calendar embed a design decision, or is there something blocking this on the technical side?

I know we've had discussions about how our current events page is less than ideal, but curious what the forcing function behind this is.

File: pages/index.js
Comment (evanzhong) [inline pages/index.js:29]: What does this comment mean?

File: styles/components/CommitteeSpread.scss
Comment (evanzhong) [inline styles/components/CommitteeSpread.scss:66]: If you don't foresee this being necessary in the future, you can delete it

File: styles/pages/committees.scss
Comment (evanzhong) [inline styles/pages/committees.scss:66]: Interesting stuff, perhaps this is what Bryan mentioned about `next/image` being too abstracted for his liking.

In the future, if this persists we may want to consider creating our own wrapper component for `next/image`

**Review by evanzhong — CHANGES_REQUESTED** [review]
Amazing work @advaithg and @mattxwang!

I mainly looked at the diffs in this review. Most of the relevant items are left in comments in specific files. General things I noticed
- +1 to the comment in the PR writeup, there's a lot of low hanging fruit with using `.modules`; a discussion we should have later after this PR ships.
- There's some extraneous commented out JSX left in files under `TODOs` comments. My personal preference is not to have commented out code in production. If we can always refer back to it in the commit history if needed down the line. I left some comments about this, but didn't comment on every instance I saw.
- We use both regular html `` tags and `next/image`. This could get confusing in future development as devs might not know which is appropriate. I know `next/image` was a particular pain point so we can talk about addressing this in a future issue.

Only thing blocking for me is removal of commented out code, other than that it looks good to go!

File: pages/index.js
Comment (mattxwang) [reply pages/index.js:29]: Good catch, I think I added it before refactoring how `.button` worked.

File: pages/events.js
Comment (mattxwang) [reply pages/events.js:33]: I believe this is actually from a previous PR I made, #173 - let me know what you think!

(forcing function - the calendar is completely empty)

File: components/Navbar.js
Comment (mattxwang) [reply components/Navbar.js:8]: Yeah, I'm not the largest fan of this either, especially since this requires people to update comments (uncommon) when implementation details change, like they did here.

The JSON implementation is a good idea; our friends at Creative Labs do something similar, see https://github.com/UCLA-Creative-Labs/sunshine/blob/master/components/Navbar.tsx#L9

File: styles/pages/committees.scss
Comment (mattxwang) [reply styles/pages/committees.scss:66]: Yup yup, honestly strange that they force so many style tag changes (causes some specificity shenanigans). Open to any suggestions on how to approach this.

File: pages/events.js
Comment (evanzhong) [reply pages/events.js:33]: sgtm, we'll probably axe it anyways when we get to #115

File: components/Committees/CommitteeEventCard.js
Comment (mattxwang) [reply components/Committees/CommitteeEventCard.js:8]: I believe so, I think this is from Advaith's side!

File: components/Carousel.js
Comment (mattxwang) [reply components/Carousel.js:26]: Yeah, so my reasoning here is that the background image approach itself is a hack - each `div` really semantically should be an image, and without this change (in general) the Carousel would be inaccessible.

I did merge that above change with *also* transitioning it to `<Image />`, which may have jumped the gun. Didn't see any issues with this on prod, but always down to switch back!

**Comment by mattxwang** [comment]
Addressed most of Evan's comments, and updated the PR description with the relevant to-dos!

File: components/Carousel.js
Comment (evanzhong) [reply components/Carousel.js:27]: Makes sense, re: accessibility.

If no problems seen on prod with `next/image` I think it's actually a great change. The carousel is a prime place for lazy loading since it's located below the fold and is image dense

**Review by evanzhong — APPROVED** [review]
lgtm for this set of changes!

**Pull request merged**
